### ISSUE TITLE
nonce field should be BytestringField

### ISSUE THREAD

**Issue by gzm55** [scheiblingco/sshkey-tools#19]
according to certkeys protocol of openssh, the nonce should be a byte string.

**Comment by scheibling** [comment]
TL;DR: That's correct, and under the hood it more or less is either way

The protocol/specification doesn't really distinguish between bytestrings/strings, a string is basically just a bytestring with encoding applied (which is how this implementation is set up as well, a string inherits from bytestring with an extra step before serialization and after deserialization to return it in string form instead).

The reason all user-customizable fields are StringFields instead of BytestringFields is mostly to avoid the confusion of bytes vs. strings for people who aren't familiar with it (since the BytestringField only takes bytes, not strings) while keeping the option to pass bytestrings (a StringField takes both) for developers who want to pass other types of data.

**Comment by gzm55** [comment]
> TL;DR: That's correct, and under the hood it more or less is either way
> 
> The protocol/specification doesn't really distinguish between bytestrings/strings, a string is basically just a bytestring with encoding applied (which is how this implementation is set up as well, a string inherits from bytestring with an extra step before serialization and after deserialization to return it in string form instead).
> 
> The reason all user-customizable fields are StringFields instead of BytestringFields is mostly to avoid the confusion of bytes vs. strings for people who aren't familiar with it (since the BytestringField only takes bytes, not strings) while keeping the option to pass bytestrings (a StringField takes both) for developers who want to pass other types of data.

Thanks for quick response. As a StringField for nonce, we cannot pass an arbitrary  bytes, e.g. random bytes. Can we add a flag for `StringField` to avoid check the encoding?

**Comment by scheibling** [comment]
Ah, yes, that's what I meant. With StringField, you can pass either a string or arbitrary bytes; with ByteStringField you can only pass bytes. Although, now that you mention it, it would become an issue when decoding/re-encoding certificates that have non-unicode characters in them, that would be a good argument for switching to a Bytestring field..

I'll look into it a bit more tomorrow when I have some time and see what alternatives there are; since the nonce field is generated automatically for 90% of use cases anyway it probably wouldn't affect much if we just switched it to a Bytestring field.

```python
# Create certificate body fields
cert_fields = CertificateFields(
    serial=1234567890,
    cert_type=1,
    key_id="someuser@somehost",
    principals=["some-user", "some-group", "production-webservers"],
    valid_after=datetime.now(),
    valid_before=datetime.now() + timedelta(hours=8),
    critical_options=[],
    extensions=[
        "permit-pty",
        "permit-X11-forwarding",
        "permit-agent-forwarding",
    ],
)
# Create certificate from existing fields
certificate = Ed25519Certificate(
    subject_pubkey=subject_pubkey,
    ca_privkey=ca_privkey,
    fields=cert_fields,
)
certificate.header.nonce = b'abcdefg'

certificate.sign()
# True
print(certificate.to_string())
# ssh-ed25519-cert-v01@openssh.com AAAAIHNzaC1lZDI1NTE5LWNlcnQtdjAxQG9wZW5zc2guY29tAAAAB2FiY2RlZmcAAAAgKKWBDsG7tlOsz+ViZ2IEQIkcJxUtJT5bRl/pJpvDjU8AAAAASZYC0gAAAAEAAAARc29tZXVzZXJAc29tZWhvc3QAAAA0AAAACXNvbWUtdXNlcgAAAApzb21lLWdyb3VwAAAAFXByb2R1Y3Rpb24td2Vic2VydmVycwAAAABkaPRNAAAAAGRpZM0AAAAAAAAATgAAAApwZXJtaXQtcHR5AAAAAAAAABVwZXJtaXQtWDExLWZvcndhcmRpbmcAAAAAAAAAF3Blcm1pdC1hZ2VudC1mb3J3YXJkaW5nAAAAAAAAAAAAAAAzAAAAC3NzaC1lZDI1NTE5AAAAIPBwJlQdNNLevEnRCQQAByPSun4llud/IvTMtCWWMzBwAAAAUwAAAAtzc2gtZWQyNTUxOQAAAEBMroxBnOW/fp6njxo8P+4lLhga/YE6R+Wr6NjGiRer+/kIzQJCJkJRp2mZaFLq2XWPoiMiOuxF0GeKiXtX6nkA 
certificate.header.nonce = 'abcdefg'
certificate.sign()
# True
print(certificate.to_string())
# ssh-ed25519-cert-v01@openssh.com AAAAIHNzaC1lZDI1NTE5LWNlcnQtdjAxQG9wZW5zc2guY29tAAAAB2FiY2RlZmcAAAAgKKWBDsG7tlOsz+ViZ2IEQIkcJxUtJT5bRl/pJpvDjU8AAAAASZYC0gAAAAEAAAARc29tZXVzZXJAc29tZWhvc3QAAAA0AAAACXNvbWUtdXNlcgAAAApzb21lLWdyb3VwAAAAFXByb2R1Y3Rpb24td2Vic2VydmVycwAAAABkaPRNAAAAAGRpZM0AAAAAAAAATgAAAApwZXJtaXQtcHR5AAAAAAAAABVwZXJtaXQtWDExLWZvcndhcmRpbmcAAAAAAAAAF3Blcm1pdC1hZ2VudC1mb3J3YXJkaW5nAAAAAAAAAAAAAAAzAAAAC3NzaC1lZDI1NTE5AAAAIPBwJlQdNNLevEnRCQQAByPSun4llud/IvTMtCWWMzBwAAAAUwAAAAtzc2gtZWQyNTUxOQAAAEBMroxBnOW/fp6njxo8P+4lLhga/YE6R+Wr6NjGiRer+/kIzQJCJkJRp2mZaFLq2XWPoiMiOuxF0GeKiXtX6nkA 
```

**Comment by gzm55** [comment]
> ```python
> timedelta
> ```
how about treat the nonce as a Bytestring, in the meanwhile, it also accept a string value, and use utf8 to encoding the string value to bytes, when do decode, just leave the raw bytes to the value.

**Comment by scheibling** [comment]
Yes, that's more or less what I'm thinking too. I'll fix it during the day

**Comment by scheibling** [comment]
The nonce field is now a Bytestring field, and I've made some minor (non-breaking) changes so the only thing that differs a StringField from a BytestringField is the default encoding and the output format when decoding a certificate.
